Add op context tracker, misc clean up

agibsonccc opened the pull request:
## What changes were proposed in this pull request?

Fix strange property name on enable array cache in nd4j constants
 Add more documentation for deallocator service
Add associations in op context for op context tracker
Add op context tracker tracking allocations and deallocations of op contexts. Mainly used to debug non workspaces related memory.

## How was this patch tested?

Ran manually

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

agibsonccc committed:
Spacing clean up
Fix strange property name on enable array cache in nd4j constants
Add more documentation for deallocator service
Add associations in op context for op context tracker

treo commented on libnd4j/CMakeLists.txt:
```diff
@@ -15,13 +15,13 @@ option(SD_CHECK_VECTORIZATION "checks for vectorization" OFF)
 option(SD_STATIC_LIB "Build static library" OFF)
 option(SD_SHARED_LIB "Build shared library" ON)
-option(SD_SANITIZE "Enable Address Sanitizer" OFF)
+option(SD_SANITIZE "Enable Address Sanitizer" ON)
```
Is this meant to be enabled by default, or is that a leftover from debugging?

treo commented on libnd4j/include/ops/declarable/impl/DeclarableOp.cpp:
```diff
@@ -1075,20 +1079,26 @@ sd::Status DeclarableOp::execute(const std::vector<NDArray *> &inputs, const std
   ctx.setTArguments(tArgs);
   ctx.setBArguments(bArgs);
   ctx.setDArguments(dArgs);
+  sd_printf("Set arguments\n",0);
```
printfs outside of error conditions.

treo commented on nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/memory/ArrayCacheMemoryMgr.java:
```diff
@@ -143,7 +143,7 @@ public static void setCurrentCacheSize(AtomicLong currentCacheSize) {
 
     private static boolean enableCache = Boolean
-            .parseBoolean(System.getProperty(ND4JSystemProperties.SAMEDIFF_MEMORY_CACHE_ENABLE, "true"));
+            .parseBoolean(System.getProperty(ND4JSystemProperties.SAMEDIFF_MEMORY_CACHE_DISABLE, "true"));
```
Why do we change this again into the disable form?

treo commented on nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/memory/deallocation/DeallocatorService.java:
```diff
@@ -37,39 +35,6 @@
- * Due to relying on the JVM to supply garbage collectable references,
- * the allocator fall behind in deallocation. This happens VERY rarely and only at certain scales.
- *
- */
```
Why do you remove that comment again?

agibsonccc replied:
Mixed around in different commits. Thanks for highlighting. Added back.

treo reviewed (changes requested):
This PR looks a bit all over the place. Why are you removing things that you added just in the last PR?

agibsonccc committed:
Update CMakeLists.txt

agibsonccc committed:
Update DeclarableOp.cpp

agibsonccc committed:
Update DeallocatorService.java

agibsonccc committed:
Spacing clean up
Fix strange property name on enable array cache in nd4j constants
Add more documentation for deallocator service
Add associations in op context for op context tracker
Fix enable cache nit

agibsonccc committed:
Add back op context changes
